tools/clang-tidy-plugin/AlmostNeverAutoCheck.cpp

```diff
@@ -0,0 +1,145 @@
+#include "AlmostNeverAutoCheck.h"
+
+#include <cassert>
+#include <clang/AST/Decl.h>
+#include <clang/AST/DeclCXX.h>
+#include <clang/AST/Expr.h>
+#include <clang/AST/ExprCXX.h>
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/ASTMatchers/ASTMatchersInternal.h>
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Basic/LLVM.h>
+#include <clang/Basic/SourceLocation.h>
+#include <clang/Lex/Lexer.h>
+#include <llvm/ADT/APInt.h>
+#include <llvm/Support/Casting.h>
+#include <map>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "Utils.h"
+#include "clang/AST/OperationKinds.h"
+#include "../../src/cata_assert.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang
+{
+namespace tidy
+{
+namespace cata
+{
+
+void AlmostNeverAutoCheck::registerMatchers( MatchFinder *Finder )
+{
+    Finder->addMatcher(
+        varDecl(
+            hasType( autoType() )
+        ).bind( "decl" ),
+        this
+    );
+}
+
+static void CheckDecl( AlmostNeverAutoCheck &Check,
+                       const MatchFinder::MatchResult &Result )
+{
+    const VarDecl *AutoVarDecl = Result.Nodes.getNodeAs<VarDecl>( "decl" );
+
+    if( !AutoVarDecl ) {
+        return;
+    }
+
+    if( AutoVarDecl->getName().startswith( "_" ) ) {
```

Comment:
Does `isImplicit` cover this? Here's the comment from `clang/ASTMatchers/ASTMatchers.h`:

```
/// Matches a declaration that has been implicitly added
/// by the compiler (eg. implicit default/copy constructors).
AST_MATCHER(Decl, isImplicit) {
  return Node.isImplicit();
}
```

Reply:
Looks like it does indeed.  Thanks, that's much nicer.